packages/ifpack2/src/Ifpack2_Details_Fic_def.hpp

```diff
@@ -0,0 +1,287 @@
+  {
+    throw std::invalid_argument("Filu::apply: X and Y have different numbers of vectors.");
+  }
+  int nvecs = X.getNumVectors();
+  ScalarArray x;
+  ScalarArray y;
+  auto x2D = X.template getLocalView<ExecSpace>();
+  auto y2D = Y.template getLocalView<ExecSpace>();
+  //Solve each vector one at a time (until FastILU supports multiple RHS)
+  for(int i = 0; i < nvecs; i++)
+  { 
+    auto x1D = Kokkos::subview(x2D, Kokkos::ALL(), i);
+    auto y1D = Kokkos::subview(y2D, Kokkos::ALL(), i);
+
+    x = x1D;
+    y = y1D;
+
+    //TODO:Implement multiple right hand sides. 
+    localPrec->apply(x, y);
+  }
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+void Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+setParameters (const Teuchos::ParameterList& List)
+{
+  //note: see compute() for the validation and defaults for parameters
+  params = List;
+  computedFlag = false;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+void Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+initialize()
+{
+  if(mat.is_null())
+  {
+    throw std::runtime_error("Called Fic::initialize() but matrix was null.");
+  }
+  Kokkos::Impl::Timer copyTimer;
+  CrsArrayReader<Scalar, LocalOrdinal, GlobalOrdinal, Node>::getRowPtrs(mat.get(), localRowPtrs);
+  CrsArrayReader<Scalar, LocalOrdinal, GlobalOrdinal, Node>::getColInds(mat.get(), localColInds);
+  crsCopyTime = copyTimer.seconds();
+  initFlag = true;
+  nInit++;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+bool Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+isInitialized() const 
+{
+  return initFlag;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+void Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+compute()
+{
+  if(!initFlag)
+  {
+    throw std::runtime_error("Fic: initialize() must be called before compute()");
+  }
+  //read parameters (set defaults for params that are not set)
+  int nFact = params.get<int>("sweeps", 1);
+  int nTrisol = params.get<int>("triangular solve iterations", 1);
+  int level = params.get<int>("level", 0);
+  double omega = params.get<double>("damping factor", 0.5);
+  double shift = params.get<double>("shift", 0);
+  bool guessFlag = params.get<bool>("guess", true);
+  int blkSz = params.get<int>("block size", 5);
+  //get copy of values array from matrix
+  Kokkos::Impl::Timer copyTimer;
+  CrsArrayReader<Scalar, LocalOrdinal, GlobalOrdinal, Node>::getValues(mat.get(), localValues);
+  crsCopyTime += copyTimer.seconds(); //add to the time spent getting rowptrs/colinds
+  auto nRows = mat->getNodeNumRows();
+  //create a new preconditioner with these params (if already exists, this replaces it)
+  localPrec = Teuchos::rcp(new LocalFIC(localRowPtrs, localColInds, localValues, nRows, nFact, nTrisol, level, omega, shift, guessFlag ? 1 : 0, blkSz));
```

Comment:
Yes, that is why. Only in compute() are the vals, rowptrs, colinds all available together. So I will add setValues() to FastILUPrec etc. in shylu and then in Ifpack2 call compute() on on localPrec_ after setting values.